tskv/src/wal.rs

```diff
@@ -105,121 +354,235 @@ impl WriteAheadLogManager {
-            .map_err(|err| Error::FileManagerError { source: err })
+            .context(FailedWithFileManagerSnafu)
     }
 
-    fn new_wal_file(&mut self) -> Result<()> {
-        self.current_file_id = self.current_file_id + 1;
-        if self.current_file_writer.is_none() {
+    fn roll_wal_file(&mut self) -> WalResult<()> {
+        // let mut current_file = self.current_file.lock().unwrap();
+        let current_file = &mut self.current_file;
+        if current_file.size > SEGMENT_SIZE {
+            current_file.id += 1;
             let file_name = self
                 .current_dir_path
-                .join(format!("_{:05}.wal", self.current_file_id));
-            self.current_file_writer = Some(
-                self.file_manager
-                    .create_file(file_name)
-                    .map_err(|err| Error::FileManagerError { source: err })
-                    .map(|file| WriteAheadLogDiskFileWriter::new(file))?,
-            );
+                .join(format!("_{:05}.wal", current_file.id));
+            let file = self
+                .file_manager
+                .create_file(file_name)
+                .context(FailedWithFileManagerSnafu)?;
+            current_file.file = Arc::new(file);
         }
         Ok(())
     }
 
-    fn roll_wal_file(&mut self) -> Result<()> {
-        let writer = &self.current_file_writer;
-        match writer {
-            Some(w) if w.size > SEGMENT_SIZE => self.new_wal_file(),
-            None => self.new_wal_file(),
-            _ => Ok(()),
-        }
-    }
+    pub async fn write(&mut self, rows: &[u8]) -> WalResult<()> {
+        self.roll_wal_file()?;
 
-    pub fn open(&mut self) -> Result<()> {
-        let segments = Self::list_wal_filenames(&self.dir);
-        if segments.len() > 0 {
-            let last = segments.last().unwrap();
-            let id = Self::get_id_by_file_name(last)?;
-            self.current_file_id = id;
-            let file = self.get_wal_file(self.current_dir_path.join(last))?;
-            self.current_file_writer = Some(WriteAheadLogDiskFileWriter::new(file));
+        // let mut writer = self.current_file.lock().unwrap();
+        // writer.writer.write(&WalEntryBlock::new_write(rows))
+
+        let wal_block = WalEntryBlockInner::from_bytes(WalEntryType::Write, rows);
+        let buf = wal_block.to_bytes();
+
+        let writer = &mut self.current_file;
+        {
+            writer
+                .file
+                .write_at(writer.size, buf.as_slice())
+                .map_err(|err| WalError::FailedWithStdIO { source: err })?;
+
+            // TODO codes below may produce "future cannot be sent between threads safely" in `kvcore.rs`."
+            // self.file_manager
+            //     .write_at(Arc::clone(&writer.file), writer.size, buf.as_mut_slice())
+            //     .await;
         }
+        writer.size += buf.len() as u64;
+        writer
+            .file
+            .sync_all(FileSync::Soft)
```

Comment:
use wal option

Reply:
will be fixed on next commit

Reply:
fixed in [09cf1f0](https://github.com/cnosdb/cnosdb/pull/237/commits/09cf1f08d3319e364443f8f512ac9a1ccdc77550)